internal/app/grpcclient/impl.go

```diff
@@ -113,24 +115,224 @@ func (i *invoker) unary(
+	payloads := make([]*v1alpha1.ConformancePayload, 0, len(req.ResponseDefinition.ResponseData))
+
+	for {
+		msg, err := stream.Recv()
+		if err != nil {
+			if !errors.Is(err, io.EOF) {
+				protoErr = grpcutil.ConvertGrpcToProtoError(err)
+			}
+			break
+		}
+		// If the call was successful, get the returned payloads
+		// and the headers and trailers
+		payloads = append(payloads, msg.Payload)
+	}
+
+	return &v1alpha1.ClientResponseResult{
+		ResponseHeaders:  headers,
+		ResponseTrailers: trailers,
+		Payloads:         payloads,
+		Error:            protoErr,
+		ConnectErrorRaw:  nil, // TODO
+	}, nil
 }
 
 func (i *invoker) clientStream(
-	_ context.Context,
-	_ *v1alpha1.ClientCompatRequest,
+	ctx context.Context,
+	ccr *v1alpha1.ClientCompatRequest,
 ) (*v1alpha1.ClientResponseResult, error) {
-	return nil, errors.New("client streaming is not yet implemented")
+	// Add the specified request headers to the request
+	ctx = grpcutil.AppendToOutgoingContext(ctx, ccr.RequestHeaders)
+
+	stream, err := i.client.ClientStream(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read headers and trailers from the stream
+	hdr, err := stream.Header()
+	if err != nil {
+		return nil, err
+	}
+	headers := grpcutil.ConvertMetadataToProtoHeader(hdr)
+	trailers := grpcutil.ConvertMetadataToProtoHeader(stream.Trailer())
+
+	for _, msg := range ccr.RequestMessages {
+		csr := &v1alpha1.ClientStreamRequest{}
+		if err := msg.UnmarshalTo(csr); err != nil {
+			return nil, err
+		}
+
+		// Sleep for any specified delay
+		time.Sleep(time.Duration(ccr.RequestDelayMs) * time.Millisecond)
+
+		if err := stream.Send(csr); err != nil && errors.Is(err, io.EOF) {
+			break
+		}
+	}
+
+	var protoErr *v1alpha1.Error
+	payloads := make([]*v1alpha1.ConformancePayload, 0, 1)
+
+	resp, err := stream.CloseAndRecv()
+	if err != nil {
+		// If an error was returned, convert it to a gRPC error
+		protoErr = grpcutil.ConvertGrpcToProtoError(err)
+	} else {
+		// If the call was successful, get the returned payloads
+		payloads = append(payloads, resp.Payload)
+	}
+
+	return &v1alpha1.ClientResponseResult{
+		ResponseHeaders:  headers,
+		ResponseTrailers: trailers,
+		Payloads:         payloads,
+		Error:            protoErr,
+		ConnectErrorRaw:  nil, // TODO
+	}, nil
 }
 
 func (i *invoker) bidiStream(
-	_ context.Context,
-	_ *v1alpha1.ClientCompatRequest,
+	ctx context.Context,
+	ccr *v1alpha1.ClientCompatRequest,
 ) (result *v1alpha1.ClientResponseResult, retErr error) {
-	return nil, errors.New("bidi streaming is not yet implemented")
+	result = &v1alpha1.ClientResponseResult{
+		ConnectErrorRaw: nil, // TODO
+	}
+
+	// Add the specified request headers to the request
+	ctx = grpcutil.AppendToOutgoingContext(ctx, ccr.RequestHeaders)
+
+	stream, err := i.client.BidiStream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Read headers and trailers from the stream
+	hdr, err := stream.Header()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
```

Comment:
Might be best off using this defer pattern above for client and server streams, too.